pl-faded-parsons.py

```diff
@@ -1,314 +1,577 @@
+def render(element_html: str, data: pl.QuestionData):
+    panel_type = data["panel"]
+
+    fpp = FadedParsonsProblem(element_html, data)
+    mustache_file = f"pl-faded-parsons-{panel_type}.mustache"
+
+    if panel_type == "question":
+        # chevron skips rendering when values are falsy (eg pre-text/post-text/starter)
+        html_params = asdict(fpp.to_mustache())
+    elif panel_type == "submission":
+        html_params = {
+            "code": fpp.to_code(),
+        }
+    elif panel_type == "answer":
+        html_params = {"solution_path": fpp.solution_path}
     else:
-        raise Exception(f'Invalid panel type: {panel_type}')
+        raise Exception(f"Invalid panel type: {panel_type}")
 
+    with open(mustache_file, "r") as f:
+        return chevron.render(f, html_params).strip()
 
-def parse(element_html, data):
-    """Parse student's submitted answer (HTML form submission)"""
-    # make an XML fragment that can be passed around to other PL functions,
-    # parsed/walked, etc
-
-    element = xml.fragment_fromstring(element_html)
-    format = pl.get_string_attrib(element, "format", "right").replace("-", '_')
-
-    def load_json_if_present(key: str, default=[]):
-        if key in data['raw_submitted_answers']:
-            return json.loads(data['raw_submitted_answers'][key])
-        return default
-
-    if format != "no_code":
-        starter_lines = load_json_if_present('starter-tray-order')
-    submission_lines = load_json_if_present('solution-tray-order')
-
-    submission_code = "\n".join([
-        line.get("content", "")
-        for line in submission_lines
-    ]) + "\n"
 
-    data['submitted_answers']['student-parsons-solution'] = submission_code
-    if format != "no_code":
-        data['submitted_answers']['starter-lines'] = starter_lines
-    data['submitted_answers']['submission-lines'] = submission_lines
+def parse(element_html: str, data: pl.QuestionData):
+    """Parse student's submitted answer (HTML form submission)"""
 
-    # `element` is now an XML data structure - see docs for LXML library at lxml.de
+    def base64_encode(s):
+        return base64.b64encode(s.encode("ascii")).decode("ascii")
 
-    # only Python problems are allowed right now (lang MUST be "py")
-    # lang = pl.get_string_attrib(element, 'language') # TODO: commenting is a stop gap for the pilot study, find a better solution
+    fpp = FadedParsonsProblem(element_html, data)
 
-    file_name = pl.get_string_attrib(element, 'file-name', 'user_code.py')
+    student_code = fpp.to_code()
 
-    data['submitted_answers']['_files'] = [
-        {
-            "name": file_name,
-            "contents": base64_encode(get_student_code(element_html, data))
-        }
-    ]
+    # provide the answer to users of pl-faded-parsons in classic PL style
+    data["submitted_answers"][fpp.answers_name] = student_code
+    pl.add_submitted_file(data, fpp.out_filename, base64_encode(student_code))
```

Comment:
This function no longer exists? Pl raises an AttributeError. Would be helpful if someone could replicate this problem because the pl source still has the function...

Reply:
I was not able to recreate this locally. Will try on codespaces